### PR TITLE
Fixed #31573 -- Made QuerySet.update() respect ordering on MariaDB/MySQL.

### PR DESCRIPTION
**Pull request by davidchorpash**
https://code.djangoproject.com/ticket/31573

Multiple rows can be updated with unique constraints on MySQL databases using QuerySet.order_by(...).update(...)

### CODE REVIEW

**Comment by davidchorpash** [comment]
I need to add a way to skip the added test for databases other than MySQL.

File: tests/update/tests.py [hunk 1]
```diff
@@ -199,3 +202,18 @@ def test_update_with_joined_field_annotation(self):
...
+    @classmethod
+    def setUpTestData(cls):
+        cls.u1 = UniqueNumber.objects.create(number=1)
+        cls.u2 = UniqueNumber.objects.create(number=2)
+
+    def test_order_by_update_on_unique_constraint(self):
+        """
+        Update a queryset using ordered_by on a unique constraint.
+        """
+        updated_count = UniqueNumber.objects.all().order_by('-number', 'id').update(number=F('number') + 1)
```
Comment (adamchainz) [inline tests/update/tests.py:218]: The `.all()` is redundant
```suggestion
        updated_count = UniqueNumber.objects.order_by('-number', 'id').update(number=F('number') + 1)
```

File: tests/update/tests.py [hunk 2]
```diff
@@ -199,3 +202,18 @@ def test_update_with_joined_field_annotation(self):
...
+@unittest.skipUnless(connection.vendor == 'mysql', 'MySQL tests')
+class MySQLUpdateTest(TestCase):
+    @classmethod
+    def setUpTestData(cls):
+        cls.u1 = UniqueNumber.objects.create(number=1)
+        cls.u2 = UniqueNumber.objects.create(number=2)
+
+    def test_order_by_update_on_unique_constraint(self):
+        """
+        Update a queryset using ordered_by on a unique constraint.
```
Comment (adamchainz) [inline tests/update/tests.py:216]:
```suggestion
        Update a queryset using order_by on a unique constraint.
```

File: django/db/backends/mysql/compiler.py [hunk 3]
```diff
@@ -36,7 +36,21 @@ def as_sql(self):
...
+
+        order_by_clause = 'ORDER BY '
+        for annotation in order_by_annotations:
+            order_by_clause += '`%s` %s' % (
+                (annotation[1:], 'DESC') if annotation[0] == '-' else (annotation, 'ASC')
+            )
+            if annotation != order_by_annotations[-1]:
+                order_by_clause += ', '
+        query_with_order_by_clause = ('%s  %s' % (query[0], order_by_clause))
+        return query_with_order_by_clause, query[1]
```
Comment (adamchainz) [inline django/db/backends/mysql/compiler.py:53]: You should use tuple deconstruction in a number of places, starting with `query, params = super().as_sql()`. See other SQLCompiler methods.

Also you seem to have replicated a bit of logic from `SQLCompiler.get_order_by()`. You shouldn't do that, but find a way to reuse it. `order_by()` supports many forms beyond the asc/desc field name form you've compiled here.

Comment (charettes) [reply django/db/backends/mysql/compiler.py:53]: > order_by() supports many forms beyond the asc/desc field name form you've compiled here.

For example, you could add a test for `order_by(F('number') * -1))`.

Comment (davidchorpash) [reply django/db/backends/mysql/compiler.py]: @adamchainz Thanks for pointing that out and I've applied these suggestions here: https://github.com/django/django/pull/13088/commits/b9f3ff953541ca2d79ad58ca4b8c4033e2006d2b

I didn't realize how much I was simplifying this, so please let me know if I should add more tests on top of the one @charettes suggested.

File: django/db/backends/mysql/compiler.py [hunk 4]
```diff
@@ -36,7 +36,12 @@ def as_sql(self):
 
 
 class SQLUpdateCompiler(compiler.SQLUpdateCompiler, SQLCompiler):
-    pass
+    def as_sql(self):
+        query, params = super().as_sql()
+        if self.query.order_by:
+            sql_list = [sql % tuple(params) for expr, (sql, params, is_ref) in self.get_order_by()]
+            query += ' ORDER BY ' + ', '.join(sql_list)
```
Comment (charettes) [inline django/db/backends/mysql/compiler.py:43]: You want to avoid interpolating params yourself as that could result in SQL injections.

```suggestion
        if self.query.order_by:
            order_by_sql = []
            order_by_params = []
            for _, (sql, params, is_ref) in self.get_order_by():
                order_by_sql.append(sql)
                order_by_params.extend(params)
            query += ' ORDER BY ' + ', '.join(order_by_sql)
            params += tuple(order_by_params)
```

Comment (charettes) [reply django/db/backends/mysql/compiler.py:43]: Also I think that `is_ref` will need to be handled in a special way. The reference will need to be resolved to inline the expression.

This can be tested by with the following code

```python
updated_count = UniqueNumber.objects.annotate(
    number_inverse=F('number') * -1
).order_by('number_inverse').update(
    number=F('number') + 2,
)
self.assertEqual(updated_count, 2)
```

In this case `number_inverse` will yield `is_ref=True` and `expr` will be an instance of `Ref` if I'm not mistaken.

Comment (davidchorpash) [reply django/db/backends/mysql/compiler.py]: I'm working on the `QuerySet.annotate().order_by().update()` use-case and I can't seem to reliably call `get_order_by()` without getting an exception since the `self.query.annotations` dict is set to empty whenever the `db.models.query.update()` is called.

As a result, I'm receiving this this exception:
`django.core.exceptions.FieldError: Cannot resolve keyword 'number_inverse' into field. Choices are: id, number`

I'm still digging into this, but I wanted to put it out there in case anyone had any ideas.

Comment (charettes) [reply django/db/backends/mysql/compiler.py]: An option would be to completely ignore such reference then. This use case is already limited and obscure to discover and since `order_by` clauses were previously silently dropped it would preserve backward compatibility?

Comment (davidchorpash) [reply django/db/backends/mysql/compiler.py]: I ended up catching FieldError exceptions in case `get_order_by()` could not be resolved.  This seems to satisfy this use-case.

File: tests/update/tests.py [hunk 5]
```diff
@@ -199,3 +202,36 @@ def test_update_with_joined_field_annotation(self):
...
+        self.assertEqual(updated_count, 2)
+
+    def test_order_by_update_on_unique_constraint_with_query_expression(self):
+        """
+        Update a queryset using an F expression as an order_by argument on a unique constraint.
+        """
+        updated_count = UniqueNumber.objects.order_by(F('number') * -1).update(number=F('number') + 2)
+        self.assertEqual(updated_count, 2)
+
+    def test_order_by_update_on_unique_constraint_with_annotation(self):
```
Comment (felixxm) [inline tests/update/tests.py:228]: This test should catch `IntegrityError`, e.g.
```python
def test_order_by_update_on_unique_constraint_annotation(self):
    # order_by() with annotate references are ignored.
    with self.assertRaises(IntegrityError):
        UniqueNumber.objects.annotate(
            number_inverse=F('number').desc(),
        ).order_by('number_inverse').update(
            number=F('number') + 2,
        )
```

File: tests/update/tests.py [hunk 6]
```diff
@@ -199,3 +202,36 @@ def test_update_with_joined_field_annotation(self):
...
+        """
+        updated_count = UniqueNumber.objects.order_by('-number', 'id').update(number=F('number') + 2)
+        self.assertEqual(updated_count, 2)
+
+    def test_order_by_update_on_unique_constraint_with_query_expression(self):
+        """
+        Update a queryset using an F expression as an order_by argument on a unique constraint.
+        """
+        updated_count = UniqueNumber.objects.order_by(F('number') * -1).update(number=F('number') + 2)
+        self.assertEqual(updated_count, 2)
```
[the remark this answers is not in the document]
Comment (davidchorpash) [reply tests/update/tests.py:226]: Thanks for the tip!  I'm used to using parametrize with pytest to get this done and this is stuff I'm unfamiliar with.

File: django/db/backends/mysql/compiler.py [hunk 7]
```diff
@@ -36,7 +38,22 @@ def as_sql(self):
...
+            try:
+                for _, (sql, params, is_ref) in self.get_order_by():
+                    order_by_sql.append(sql)
+                    order_by_params.extend(params)
+                update_query += ' ORDER BY ' + ', '.join(order_by_sql)
+                update_params += tuple(order_by_params)
+            except FieldError:
+                # Certain cases can not resolve self.get_order_by().
+                # If this is the case, we will throw an IntegrityError.
+                raise IntegrityError
```
Comment (felixxm) [inline django/db/backends/mysql/compiler.py:55]: You don't need to raise an `IntegrityError`, database will do this for us if necessary.

Comment (felixxm) [inline django/db/backends/mysql/compiler.py:55]:
```suggestion
                pass
```

Comment (davidchorpash) [reply django/db/backends/mysql/compiler.py:55]: To test if `IntegrityError` has been raised in this [suggestion](https://github.com/django/django/pull/13088#discussion_r448289837), I felt this would be the way to do it.  Otherwise, that particular test fails.  

I can revert that change back to `pass` when handling the `FieldError`, but I'm unsure how to follow through with the suggestion linked above regarding the test.

Comment (felixxm) [reply django/db/backends/mysql/compiler.py:55]: `IntegrityError` will be raised by a database, because `ORDER BY` clause is omitted in this case. We don't need to do this manually.

Comment (davidchorpash) [reply django/db/backends/mysql/compiler.py]: It seems to accepting the query and not throwing the `IntegrityError` without the `ORDER_BY` clause.

https://djangoci.com/job/pull-requests-bionic/database=mysql,label=bionic-pr,python=python3.8/6939/testReport/junit/update.tests/MySQLUpdateOrderByTest/test_order_by_update_on_unique_constraint_with_annotation/

Is there something that's missing or off about `test_order_by_update_on_unique_constraint_with_annotation` that should be causing an `IntegrityError` to be raised?

File: tests/update/tests.py [hunk 8]
```diff
@@ -199,3 +202,38 @@ def test_update_with_joined_field_annotation(self):
...
+    def test_order_by_update_on_unique_constraint_with_annotation(self):
+        """
+        Updating a queryset using order_by with an F expression argument on an annotation
+        will raise an IntegrityError.
+        """
+        with self.assertRaises(IntegrityError):
+            UniqueNumber.objects.annotate(
+                number_inverse=F('number') * -1
+            ).order_by('number_inverse').update(
+                number=F('number') + 2,
```
Comment (felixxm) [inline tests/update/tests.py:238]: It should be `number=F('number') + 1`.

File: django/db/backends/mysql/compiler.py (hunk 7 above)
Comment (felixxm) [reply django/db/backends/mysql/compiler.py]: Yes, see https://github.com/django/django/pull/13088/files#r450658069.

Comment (davidchorpash) [reply django/db/backends/mysql/compiler.py]: Ah, thank you for pointing that out! :)

File: tests/update/tests.py [hunk 9]
```diff
@@ -199,3 +202,38 @@ def test_update_with_joined_field_annotation(self):
...
+    """Update field with a unique constraint using an ordered queryset."""
+    @classmethod
+    def setUpTestData(cls):
+        UniqueNumber.objects.create(number=1)
+        UniqueNumber.objects.create(number=2)
+
+    def test_order_by_update_on_unique_constraint(self):
+        """
+        Update a queryset using order_by on a unique constraint.
+        """
```
Comment (felixxm) [inline tests/update/tests.py:218]: We can chop this docstring.

File: tests/update/tests.py [hunk 10]
```diff
@@ -199,3 +202,38 @@ def test_update_with_joined_field_annotation(self):
...
+        for ordering in tests:
+            with self.subTest(ordering=ordering):
+                updated_count = UniqueNumber.objects.order_by(*ordering).update(number=F('number') + 1)
+                self.assertEqual(updated_count, 2)
+
+    def test_order_by_update_on_unique_constraint_with_annotation(self):
+        """
+        Updating a queryset using order_by with an F expression argument on an annotation
+        will raise an IntegrityError.
+        """
```
Comment (felixxm) [inline tests/update/tests.py:233]: https://github.com/django/django/pull/13088#discussion_r448277519

**Comment by gabn88** [comment]
Wow, would love to see this merged. Needed it again today. Thanks for the quick PR!

**Comment by charettes** [comment]
@gabn88 not sure I'm following you here.

From what I can see [using `FOR UPDATE` in `UPDATE` queries is disallowed](https://dev.mysql.com/doc/refman/8.0/en/update.html#idm45705192946032) and this ticket is only about preserving ordering when performing an `UPDATE` query.

Also, from what I know, [`FOR UPDATE` doesn't allow specifying column names but only aliases to table included in the query from its `FOR` option](https://dev.mysql.com/doc/refman/8.0/en/select.html#idm45705194357888).

In other words calling `.update` will continue to ignore previous `.select_for_update()` calls after this patch is merged.

**Comment by gabn88** [comment]
@charettes It seems I was doing something not supported by mysql and I can see why it is unrelated. Thanks for your answer!

File: tests/update/tests.py (hunk 9 above)
Comment (davidchorpash) [reply tests/update/tests.py:218]: Meaning we can get rid of it?

**Comment by davidchorpash** [comment]
> @davidchorpash Thanks for updates 👍 Only release notes and a short paragraph in docs are missing.

@felixxm Thanks for all the feedback!  I plan to work on updating the docs today.

File: tests/update/tests.py (hunk 9 above)
Comment (felixxm) [reply tests/update/tests.py:218]: Yes.

**Comment by felixxm** [comment]
@davidchorpash Thanks :+1: Welcome aboard :boat: 

I added release notes and pushed edits to docs.

File: docs/ref/models/querysets.txt [hunk 11]
```diff
@@ -2530,6 +2530,21 @@ update a bunch of records for a model that has a custom
...
+
+Chaining ``order_by()`` with ``update()`` is supported only on MariaDB and
+MySQL, and is ignored for different databases. This is useful for updating a
+unique field in the order that is specified. For example::
+
+    Entry.objects.order_by('-number').update(number=F('number') + 1)
+
+.. note::
+
+    The ``order_by()`` clause with annotations is always ignored.
```
Comment (adamchainz) [inline docs/ref/models/querysets.txt:2546]:
```suggestion
    If the ``order_by()`` clause contains annotations, it will be ignored.
```

File: docs/ref/models/querysets.txt [hunk 12]
```diff
@@ -2530,6 +2530,21 @@ update a bunch of records for a model that has a custom
...
         e.save()
 
+Ordered queryset
+^^^^^^^^^^^^^^^^
+
+.. versionadded:: 3.2
+
+Chaining ``order_by()`` with ``update()`` is supported only on MariaDB and
+MySQL, and is ignored for different databases. This is useful for updating a
+unique field in the order that is specified. For example::
```
Comment (adamchainz) [inline docs/ref/models/querysets.txt:2540]:
```suggestion
unique field in the order that is specified without conflicts. For example::
```

**Comment by davidchorpash** [comment]
Thanks @felixxm and @adamchainz!
